### PR TITLE
DNM: notes/questions

### PR DESCRIPTION
**Pull request by jgravois**

### CODE REVIEW

File: src/utilities/object-mapper/typescript-object.ts [hunk 1]
```diff
@@ -37,6 +41,10 @@ export class TypeScriptObjectMap<
   }
 
   protected metaTypeData(type: GraphQLType, metaType: MetaTypeData = {}): MetaTypeData {
+    const thisWorksToo = MoreTypeGuards.isScalar(type)
+
+    if (thisWorksToo) type
```
Comment (jgravois) [inline src/utilities/object-mapper/typescript-object.ts:46]: Screenshot 2024-10-17 at 11 11 35 AM

**Review by jgravois — COMMENTED** [review]
besides `npm install` what do i need to get this up and running locally?

```
➜  ts-gql git:(qs) npm start

> ts-gql@1.0.0 start
> node dist/bin.js

node:internal/modules/cjs/loader:1147
  throw err;
  ^

Error: Cannot find module '/Users/john/code/ts-gql/dist/bin.js'
    at Module._resolveFilename (node:internal/modules/cjs/loader:1144:15)
    at Module._load (node:internal/modules/cjs/loader:985:27)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:135:12)
    at node:internal/main/run_main_module:28:49 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}

Node.js v20.11.0
```

File: src/utilities/object-mapper/base.ts [hunk 2]
```diff
@@ -74,6 +74,8 @@ export abstract class BaseObjectMap<T> {
   }
 
   protected findBaseType<T extends GraphQLType | GraphQLField<any, any>>(type: T): GraphQLType {
+    // this 'in' isnt familiar syntax to me.
+    // can you explain what it does?
```
Comment (Emily-Waters) [inline src/utilities/object-mapper/base.ts:78]: Here's the javascript implementation and usage.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/in

Here's a brief explainer for how it works in typescript

https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-guards-and-differentiating-types

Essentially, in typescript, the 'in' operator can be used for type discrimination when working with unions. Specifically, in this case a 'type' may have either a 'type' or 'ofType' field but not both. Without discriminating the unions, typescript will throw an error if trying to access the property directly (```type.type``` or ```type.ofType``` will error in this case)

File: src/guards/type-guards-alt.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import { GraphQLScalarType } from "graphql";
+
+// no wrapping class. function returns a plain ol boolean
+export function isScalar(type: unknown): boolean {
```
Comment (Emily-Waters) [inline src/guards/type-guards-alt.ts:4]: this generally works, but for more safety, type predicates are the way to go

https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates

File: src/index.ts [hunk 4]
```diff
@@ -15,6 +15,7 @@ export type Config = {
   options?: ConfigOptions;
 };
 
+// why just pass the argument directly back?
```
Comment (Emily-Waters) [inline src/index.ts:18]: in the config file, the user could define the needed export object by declaring it and then enforcing the type by importing the ```Config``` type, but I prefer exporting the defineConfig function as the default export to mark it clearly as the preferred interface

File: package.json [hunk 5]
```diff
@@ -13,7 +13,7 @@
     "test": "node --inspect-brk dist/bin.js"
   },
   "keywords": [],
-  "author": "",
+  "author": "from boilerplate? from scratch?",
```
Comment (Emily-Waters) [inline package.json:16]: scratch

File: .prettierrc [hunk 6]
```diff
@@ -1,5 +1,5 @@
 {
-  "arrowParens": "always",
+  "arrowParens": "always",  // to wrap implicit return values in parentheses?
```
Comment (Emily-Waters) [inline .prettierrc:2]:
```
const exampleToLower = str => str.lowerCase();

const exampleConcat = (a,b) => a+ b();
```

This rule is to enfore the parentheses around the function args, it's a stylistic choice but I find that it increases readability and adds consistency. Any arrow function with more than one arg requires the parentheses, this rule enforces is for arrow functions with only a single arg as well

File: src/guards/type-guards-alt.ts (hunk 3 above)
Comment (jgravois) [reply src/guards/type-guards-alt.ts:4]: yes, but my understanding is that `instanceof` is itself a type predicate and you're just wrapping it in _another_ one. 😄

File: .prettierrc (hunk 6 above)
Comment (jgravois) [reply .prettierrc:2]: ah, yes. i'm into it.

**Pull request closed without merging**